File: 백준/Silver/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.py
from collections import deque
import sys 
input = sys.stdin.readline
def bfs():
    queue = deque()
    queue.append([home[0], home[1]])
    while queue:
        x, y = queue.popleft()
        if abs(x - rock[0]) + abs(y - rock[1]) <= 1000: #맨하탄거리 / 맥주20병*50m
            return "happy"
        for i in range(store_cnt):
            if not visited[i]: #n번째 편의점 방문X
                nx, ny = store[i]
                if abs(x - nx) + abs(y - ny) <= 1000: #맥주 소진 전까지 다음 편의점 방문 가능
                    queue.append([nx, ny])
                    visited[i] = True
    return "sad"
tc = int(input())
for _ in range(tc):
    store_cnt = int(input())
    home = list(map(int, input().split()))
    store=[list(map(int, input().split())) for _ in range(store_cnt)]
    rock = list(map(int, input().split()))
    visited = [False for _ in range(store_cnt+1)] #home 제외
    print(bfs())

File: 백준/Silver/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

type Queue struct {
	X []float64
	Y []float64
}

func (q *Queue) Enqueue(x, y float64) {
	q.X = append(q.X, x)
	q.Y = append(q.Y, y)
}
func (q *Queue) Dequeue() (error, x float64, y float64) {
	if len(q.X) <= 0 {
		return 1, -1, -1
	} else {
		x = q.X[0]
		y = q.Y[0]
		if len(q.X) == 0 {
			q.X = []float64{}
			q.Y = []float64{}
		} else {
			q.X = q.X[1:]
			q.Y = q.Y[1:]
		}
		return 0, x, y
	}
}
func (q *Queue) Len() int {
	return len(q.X)
}

func bfs(home []float64, store [][]float64, rock []float64) string {
	var visited = make([]bool, len(store))
	var queue = new(Queue)
	var ret = "sad"
	var nx, ny float64
	queue.Enqueue(home[0], home[1])
	for queue.Len() > 0 {
		e, x, y := queue.Dequeue()
		if e > 0 {
			break
		}
		if math.Abs(x-rock[0])+math.Abs(y-rock[1]) <= 1000 {
			ret = "happy"
			break
		}
		for i := 0; i < len(store); i++ {
			if !visited[i] {
				nx, ny = store[i][0], store[i][1]
				if math.Abs(x-nx)+math.Abs(y-ny) <= 1000 {
					queue.Enqueue(nx, ny)
					visited[i] = true
				}
			}
		}
	}
	return ret
}
func main() {
	reader := bufio.NewReader(os.Stdin)
	var tc, storeCnt int
	var home = make([]float64, 2)
	var rock = make([]float64, 2)
	fmt.Fscanf(reader, "%d\n", &tc)
	for i := 0; i < tc; i++ {
		fmt.Fscanf(reader, "%d\n%f %f\n", &storeCnt, &home[0], &home[1])
		var store = make([][]float64, storeCnt)
		for j := 0; j < storeCnt; j++ {
			store[j] = make([]float64, 2)
			fmt.Fscanf(reader, "%f %f\n", &store[j][0], &store[j][1])
		}
		fmt.Fscanf(reader, "%f %f\n", &rock[0], &rock[1])
		fmt.Println(bfs(home, store, rock))
	}
}
